File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int leftMax = height[left];
        int rightMax = height[right];

        int water = 0;
        while(left < right) {
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);

            if(leftMax < rightMax) {
                water += (leftMax - height[left]);
                left++;
            }
            else {
                water += (rightMax - height[right]);
                right--;
            }
        }

        return water;
    }
}

/*
    좌/우 기둥 높이에서 짧은 쪽 기준으로 물이 참
    짧은 쪽 기준으로 차는데 현재에 벽이 있을 수 있으므로 현재 높이를 또 뺴주면 됨
    
    기둥 중에서 젤 높은애가 있으면 걔를 기준으로 계속 비교가 될거임 
    lmax, rmax 계속 비교 하는데
    같으면 어느쪽을 땡겨도 상관없음

    minHeight = min(lmax, rmax)
    water[i] = minHeight(lmax or rmax) - curHeight[i]
    
    만약 lmax < rmax 면 minHeight는 lmax이므로 결국
    if(lmax < rmax) {
        water[i] = lmax - height[i]
        left++;
    } else {
        water[i] = rmax - height[i]
        right--;
    }

    
*/

File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.ts
function trap(height: number[]): number {
    
    let left = 0;
    let right = height.length - 1;
    let lmax = height[left];
    let rmax = height[right];

    let water = 0;
    while(left < right) {
        lmax = Math.max(height[left], lmax);
        rmax = Math.max(height[right], rmax);

        if(lmax < rmax) {
            water += (lmax - height[left]);
            left++;
        } else {
            water += (rmax - height[right]);
            right--;
        }
    }

    return water;
};


/*
    얘는 Container With Rain Water랑 좀 다른게 
    위 문제는 양 끝 막대사이 차오르는 물 넓이를 구하는거고 

    얘는 각 칸마다 물이 차는데 그거의 누적을 구하는거임 
    그래서 해당 칸마다 왼/오 높이를 계속 비교해서 누가 더 긴지 짧은지 대봐야함 
    루프마다 lmax, rmax 구해준다음 물 계산해야됨
    그리고 width 개념 없이 그냥 높이의 차이만큼만 물이 차는거임

    각 칸마다 [왼쪽최대, 오른쪽최대 중에서 짧은 높이] 를 알아야함 
    물은 칸 단위로 누적됨 
    leftmax, rightmax를 계속 추적해야함
    

*/